### PR TITLE
cus-3267 Updated BANNO.LOANPAYMENT.SKIP.V1.POW Loan List JSon ineligibilityReason message

### PR DESCRIPTION
**Pull request by jordan-keenan**
Updated to Loan List JSon ineligibilityReason message include ineligible loan type, warning codes & service codes.

Per the following JIRA request:
Description
Edit Description
On the Skip a Pay PowerOn, update the ineligibilityReason to include the ineligible loan type, loan service code, and warning code numbers.  This will improve consistency across our PowerOns and make it easier to find when working support cases.

Ineligible loan types

Ineligible loan service codes

Ineligible account and loan warning codes

For Clarification:
Per an email from Tiffany today, 6/24/2024, she wants the words changed in Error message “Required Service Codes(s) not Found” changed to reflect Service code Eligibility, and to show which Service codes are actually on the account:

Tiffany:

“To keep it consistent and not make it too wordy, I’d rather keep the CFG setting out and change the wording to specify ineligible service code and list the service code(s) the loan has, rather than stating the Required Service code(s) not found.“

Here is what I did based on that email I got from Tiffany.

I changed it to “Eligible Service Code(s) not found. It has Code(s): “

.

[Jira CUS-3267](https://banno-jha.atlassian.net/browse/CUS-3267)

Note: Here is how I verified that the Json length was not going over 132 in the output. I added debug code which I've listed at the bottom of this:

I put in some debug code for the "Account Warning Found" and got the output for the longest Possible message 127 characters (which is less than 132) after the NEWLINE and before any other NEWLINE Command. Am I good to go?
Here is that output
{
&quot;name&quot;: &quot;USED AUTO/TRUCK UNDER 5 YR&quot;, &quot;accountId&quot;:&quot;0005&quot;,
&quot;ineligibilityReason&quot;: &quot;Account Warning Found: 001,002,003,005,006,008,009,010,011,012,013,014,015,016,017,018,019,020,021,102&quot;
,LENGTH OF PREVIOUS ineligibilityReason is            127

}

I tested the same thing for "Loan Warning Found" and got the output for the longest Possible message 120 characters (which is less than 132) after the NEWLINE and before any other NEWLINE Command.

Here is that output
,{
&quot;name&quot;: &quot;LINE OF CREDIT&quot;,
&quot;accountId&quot;:&quot;00LL&quot;,
&quot;ineligibilityReason&quot;: &quot;Loan Warning Found: 003,004,005,006,007,008,009,010,011,012,013,014,015,016,017,018,019,020,021&quot;
,LENGTH OF PREVIOUS ineligibilityReason is            120

}

I tested the same thing for “Eligible Service Code(s) not found. It has Code(s): “ and got the output for the longest Possible message 100 characters (which is less than 132) after the NEWLINE and before any other NEWLINE Command.

Here is that output
,{
&quot;name&quot;: &quot;LINE OF CREDIT&quot;,
&quot;accountId&quot;:&quot;00LL&quot;,
&quot;ineligibilityReason&quot;: &quot;Eligible Service Code(s) not found. It has Code(s): 01,03,04,05,06,07,08,09&quot;
,LENGTH OF PREVIOUS ineligibilityReason is            100

}

PRINT Q+"ineligibilityReason"+Q+": "+Q+INELIGIBLEREASONS(TEMPLOOP)
         IF TEMPLOOP=INELIGIBLECODELOANTYPE THEN
          DO
           PRINT ": "+INVLNTYPE(LOANLOOP)
           PRINT Q
          END
         ELSE IF TEMPLOOP=INELIGIBLECODEACCTWARNING THEN
          DO
           PRINT ": "+INVACCTWARNS(LOANLOOP)
           PRINT Q
[JPK TEMP DEBUG]
[PRINT Q+"ineligibilityReason"+Q+": "+Q+INELIGIBLEREASONS(TEMPLOOP)] [ THIS DEBUG CODE CURRENTLY TURNED ON ]
MESSAGELENGTHTOCHECK=Q+"ineligibilityReason"+Q+": "+Q+INELIGIBLEREASONS(TEMPLOOP)+": "+INVACCTWARNS(LOANLOOP)+Q NEWLINE
  PRINT ","
  PRINT "LENGTH OF PREVIOUS ineligibilityReason is "+FORMAT("#############9",LENGTH(MESSAGELENGTHTOCHECK))
  NEWLINE
[END JPK TEMP DEBUG]
          END
         ELSE IF TEMPLOOP=INELIGIBLECODELOANWARNING THEN
          DO
           PRINT ": "+INVLNWARNS(LOANLOOP)
           PRINT Q
[JPK TEMP DEBUG]
[PRINT Q+"ineligibilityReason"+Q+": "+Q+INELIGIBLEREASONS(TEMPLOOP)] [ THIS DEBUG CODE CURRENTLY TURNED ON ]
MESSAGELENGTHTOCHECK=Q+"ineligibilityReason"+Q+": "+Q+INELIGIBLEREASONS(TEMPLOOP)+": "+INVLNWARNS(LOANLOOP)+Q NEWLINE
  PRINT ","
  PRINT "LENGTH OF PREVIOUS ineligibilityReason is "+FORMAT("#############9",LENGTH(MESSAGELENGTHTOCHECK))
  NEWLINE
[END JPK TEMP DEBUG]
          END
         ELSE IF TEMPLOOP=INELIGIBLECODESERVICECODE THEN
          DO
           PRINT ". It has Code(s): "+EXISTINGLNSERVS(LOANLOOP)
           PRINT Q
[JPK TEMP DEBUG]
[PRINT Q+"ineligibilityReason"+Q+": "+Q+INELIGIBLEREASONS(TEMPLOOP)] [ THIS DEBUG CODE CURRENTLY TURNED ON ]
MESSAGELENGTHTOCHECK=Q+"ineligibilityReason"+Q+": "+Q+INELIGIBLEREASONS(TEMPLOOP)+". It has Code(s): "+EXISTINGLNSERVS(LOANLOOP)+Q NEWLINE
  PRINT ","
  PRINT "LENGTH OF PREVIOUS ineligibilityReason is "+FORMAT("#############9",LENGTH(MESSAGELENGTHTOCHECK))
  NEWLINE
[END JPK TEMP DEBUG]
          END
         ELSE
          PRINT Q
         NEWLINE

### CODE REVIEW

**Review by tkainz — APPROVED** [review]
Approving as-is, although there are a few things you may want to consider tweaking.  Thanks.

File: loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW [hunk 1]
```diff
@@ -429,7 +437,16 @@ PROCEDURE PROCGETPRELOADDATA
...
-         PRINT Q+"ineligibilityReason"+Q+": "+Q+INELIGIBLEREASONS(TEMPLOOP)+Q
+         PRINT Q+"ineligibilityReason"+Q+": "+Q+INELIGIBLEREASONS(TEMPLOOP)
+         IF TEMPLOOP=INELIGIBLECODELOANTYPE THEN
+          PRINT ": "+INVLNTYPE(LOANLOOP)
+         ELSE IF TEMPLOOP=INELIGIBLECODEACCTWARNING THEN
+          PRINT ": "+INVACCTWARNS(LOANLOOP)
+         ELSE IF TEMPLOOP=INELIGIBLECODELOANWARNING THEN
+          PRINT ": "+INVLNWARNS(LOANLOOP)
+         ELSE IF TEMPLOOP=INELIGIBLECODESERVICECODE THEN
+          PRINT ". It has Code(s): "+EXISTINGLNSERVS(LOANLOOP)
```
[the remark this answers is not in the document]
Comment (jordan-keenan) [reply loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW:448]: Thanks, Tom. That makes more sense. I've moved it and the others to the reason description section defined in the initialization procedure. Will be in the next commit

File: loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW [hunk 2]
```diff
@@ -429,7 +437,16 @@ PROCEDURE PROCGETPRELOADDATA
...
+         PRINT Q
          NEWLINE
          INELIGIBLEDISPLAYED=TRUE
         END
```
[the remark this answers is not in the document]
Comment (jordan-keenan) [reply loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW:452]: Hi @Tom, Its a regular For Loop that increments to a max number. UNTIL is for FOR EACH loops.

So I need to keep it as is.
It looks like the the person who programmed this procedure, did the equivalent of UNTIL INELIGIBLEDISPLAYED=TRUE, by putting a condition at the top that checks to make sure INELIGIBLEDISPLAYED=FALSE before further evaluation on each loop.

     FOR TEMPLOOP=1 TO INELIGIBLEREASONSMAX
      DO
       IF LOANINELIGIBLECODE(LOANLOOP,TEMPLOOP)=TRUE AND
          INELIGIBLEDISPLAYED=FALSE THEN
        DO
         PRINT Q+"ineligibilityReason"+Q+": "+Q+INELIGIBLEREASONS(TEMPLOOP)
         IF TEMPLOOP=INELIGIBLECODELOANTYPE THEN
          PRINT INVLNTYPE(LOANLOOP)
         ELSE IF TEMPLOOP=INELIGIBLECODEACCTWARNING THEN
          PRINT INVACCTWARNS(LOANLOOP)
         ELSE IF TEMPLOOP=INELIGIBLECODELOANWARNING THEN
          PRINT INVLNWARNS(LOANLOOP)
         ELSE IF TEMPLOOP=INELIGIBLECODESERVICECODE THEN
          PRINT EXISTINGLNSERVS(LOANLOOP)
         PRINT Q
         NEWLINE
         INELIGIBLEDISPLAYED=TRUE
        END
      END

File: loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW [hunk 3]
```diff
@@ -138,10 +142,14 @@ DEFINE
  LOANTYPES           = NUMBER        ARRAY(9999)
  SHARETYPES          = NUMBER        ARRAY(9999)
  TYPESMAX            = 9999
+ INVLNTYPE           = CHARACTER(4)  ARRAY(25)
```
[the remark this answers is not in the document]
Comment (jordan-keenan) [reply loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW:145]: Made the change. It will be in the next commit

File: loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW (hunk 1 above)
[the remark this answers is not in the document]
Comment (rrobison44) [reply loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW:448]: Not sure I love the "It has Code(s): " verbiage as part of the main error message.  I feel like it should be part of what we have been calling the "modifier".....but I see what we are trying to do here.  Following the pattern, maybe the error with modifier could be something like:
```
Eligible Service Code(s) not found: existing service codes = 10,20,25
```

**Review by rrobison44 — CHANGES_REQUESTED** [review]
See comments.  

One thing to consider is that these ineligibilityReason messages are actually displayed to the end-user.  I assume that is the intention but this is a little different than a similar thing we have done with the error messages -- which are not displayed in the UX.

Also, I thought that we had reworked most of the error sections of the json contracts to be more consistent.  Seems like this one still has some of the old formatting.  Might want to check the custom version of loan payment skip for Mountain Crest (704) and integrate similar changes into the baseline version.  Example:
image

[Mountain Crest json contract](https://github.com/Banno/custom-powerons/blob/main/704-MountainCrest%20CU/15033028%20-%20loan-payment-skip/loan-payment-skip-json-contract.md)

File: loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW [hunk 4]
```diff
@@ -1252,9 +1294,13 @@ PROCEDURE GETLOANDATA
         SERVICEFOUND=FALSE
       END
     END
-[ If none found, and some are defined]
+[* If none found, and some are defined set ineligible code.
+** Else eligible, so clear unneeded EXISTINGLNSERVS(TOTALLOANS)
+*]
    IF SERVICEFOUND=FALSE THEN
     LOANINELIGIBLECODE(TOTALLOANS,INELIGIBLECODESERVICECODE)=TRUE
+   ELSE
```
[the remark this answers is not in the document]
Comment (jordan-keenan) [reply loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW]: Hi @rrobison44 , Thank you for the good Suggested changes.  I've made your suggested changes. Can you review my second commit now?

**Review by rrobison44 — APPROVED** [review]

**Comment by jordan-keenan** [comment]
> Changes look good Jordan!

Thanks for the code review @jucarson

File: loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW [hunk 5]
```diff
@@ -437,17 +438,31 @@ PROCEDURE PROCGETPRELOADDATA
...
-          PRINT INVLNTYPE(LOANLOOP)
-         ELSE IF TEMPLOOP=INELIGIBLECODEACCTWARNING THEN
-          PRINT INVACCTWARNS(LOANLOOP)
-         ELSE IF TEMPLOOP=INELIGIBLECODELOANWARNING THEN
-          PRINT INVLNWARNS(LOANLOOP)
-         ELSE IF TEMPLOOP=INELIGIBLECODESERVICECODE THEN
-          PRINT EXISTINGLNSERVS(LOANLOOP)
-         PRINT Q
-         NEWLINE
+         IF TEMPLOOP=INELIGIBLECODELOANTYPE OR
```
Comment (rrobison44) [inline loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW:441]: I think you can simplify this quite a bit.  Something like this:
```
         INELIGDETAIL=""
         IF TEMPLOOP=INELIGIBLECODELOANTYPE THEN
          INELIGDETAIL=INVLNTYPE(LOANLOOP)
         ELSE IF TEMPLOOP=INELIGIBLECODEACCTWARNING THEN
          INELIGDETAIL=INVACCTWARNS(LOANLOOP)
         ELSE IF TEMPLOOP=INELIGIBLECODELOANWARNING THEN
          INELIGDETAIL=INVLNWARNS(LOANLOOP)
         ELSE IF TEMPLOOP=INELIGIBLECODESERVICECODE THEN
          INELIGDETAIL=="Existing service codes = "+EXISTINGLNSERVS(LOANLOOP)

         PRINT Q+"ineligibilityReason"+Q+": "+Q+INELIGIBLEREASONS(TEMPLOOP)+Q+","
         NEWLINE
         PRINT Q+"ineligibilityDetail"+Q+": "+Q+INELIGDETAIL+Q
         NEWLINE
         INELIGIBLEDISPLAYED=TRUE
```

Comment (rrobison44) [reply loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW:441]: This removes some duplicate code.  It is fine to just have an empty `ineligibilityDetail` for those reasons that don't have a detail.

**Review by rrobison44 — CHANGES_REQUESTED** [review]
See comments.  Also, I would say you should also update the contract since `ineligibilityDetail` is now being returned.  The UX won't use it for anything but I think it is still a good idea to include it in the contract.

**Review by tkainz — APPROVED** [review]
Not finding anything else that Robby hasn't caught!

File: loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW (hunk 5 above)
Comment (jordan-keenan) [reply loan-payment-skip/REPWRITERSPECS/BANNO.LOANPAYMENT.SKIP.V1.POW]: Thanks for cleaning that up @rrobison44 . I've made the change and tested that it work. How does the code look now?

**Comment by jordan-keenan** [comment]
> Not finding anything else that Robby hasn't caught!

Thank you Tom

**Comment by jordan-keenan** [comment]
> Changes look good Jordan!

Thank you Justin

**Review by rrobison44 — APPROVED** [review]
Latest changes look good.
